### DISCUSSION TITLE
Email should be optional

### DISCUSSION THREAD

**Discussion by ghost** [dani-garcia/vaultwarden#2783] [Q&A]
I suggest changing the requirement of using an email address with each account to be optional. My system has no email setup, and I imagine most people are in the same boat, so forcing all accounts to have an email as a username and logging in with it doesn't make sense. In this situation it's just a longer username ending in a domain name which may or may not even be a valid email address.

My suggestion:
Allow users to log in with a username rather than an email address, and for those who have one and the server is set up to allow for it, allow them to log in with both their email address and their username.

I suggest this here rather than the official Bitwarden page, as I don't see them having any other choice to require an email with their cloud setup. For a privately hosted setup, there is no pressing need for email, so it should be optional and up to the admin to decide.

**Comment by BlackDex** [comment]
You need to provide a valid email format as username, other than that email is not required for anything. You don't even need to configure an SMTP server.

You could even use `fake-user@fakedomain.tld` of you want as username.

**Reply by ghost** [reply]
Bitwarden relies on the email given to be a working email. One major event where this pops up is when you want to change your email. It needs to be verified with a link sent to the email address. Entering a non-existent email address brings up this issue and in this case adds unnecessary clutter to remember and type upon login, which is itself unnecessary if the server can't send emails to begin with.

It could be made so
- Email structure has no checking, as in doesn't check for an @ sign or .tld and will accept anything
- Changing emails require no verification

**Reply by BlackDex** [reply]
Are you sure you are taking about Vaultwarden??

Regarding the format of the username, that is not something we can easily change, since that is coded into all the clients, and the Vaultwarden maintainers do not maintain the clients.

I could be wrong from the top off my head, but if SMTP is not enabled, you do not need to verify a change.

**Reply by stefan0xC** [reply]
> I could be wrong from the top off my head, but if SMTP is not enabled, you do not need to verify a change.

The web-vault asks for a verification code.

We could skip this step in vaultwarden  (by immediately changing the `user.email` in an else branch in [post_email_token](https://github.com/dani-garcia/vaultwarden/blob/6fa6eb18e88f6d090eff5139d5a67bb56d25bf7f/src/api/core/accounts.rs#L418)) when mail is disabled but the change would probably be without a notification in the client. (And I am not sure if throwing an error to indicate that the change has happened would be acceptable here.)

**Reply by BlackDex** [reply]
If I read the code correctly it does exactly what i thought it would do.
It doesn't check the token if mail is disabled. You could in theory enter any code you want and it would be valid.

I'm not sure if we can somehow mention this somewhere on screen, because this is hardcoded into the web-vault, and having different use cases for this depending on a Vaultwarden specific config is going to be difficult.

Changing it directly and respond with wat error message is wrong also, that will not trigger the web-vault to run other needed actions.

Best would be if we are able to return a non error message which explains to just enter anything and continue. Not sure if we are able to do that though.
